staking/staking_hub.go

```diff
@@ -0,0 +1,394 @@
+var stakingHubInstance StakingHub
+var once sync.Once
+
+// GetStakingHub initializes the stakingHubInstance singleton
+func GetStakingHub() *StakingHub {
+	once.Do(func() {
+		stakingHubInstance = StakingHub{
+			StakingMap: make(map[types.Address]*big.Int),
+			EventQueue: make([]PendingEvent, 0),
+			CloseCh:    make(chan struct{}),
+		}
+	})
+
+	return &stakingHubInstance
+}
+
+// SetWorkingDirectory sets the writeback directory for the staking map
+func (sh *StakingHub) SetWorkingDirectory(directory string) {
+	sh.WorkingDirectory = directory
+
+	err := sh.readFromDisk()
+	if err != nil {
+		// Log as warning because reading this map is not crucial
+		sh.log(err.Error(), logWarning)
+	}
+
+	go sh.saveToDisk()
+}
+
+// LogType defines possible log types for the logger
+type LogType string
+
+var (
+	logInfo    LogType = "info"
+	logError   LogType = "error"
+	logWarning LogType = "warning"
+)
+
+// log logs the output to the console if the logger is set
+func (sh *StakingHub) log(output string, logType LogType) {
+	if sh.Logger != nil {
+		switch logType {
+		case logInfo:
+			sh.Logger.Info(output)
+		case logError:
+			sh.Logger.Error(output)
+		case logWarning:
+			sh.Logger.Warn(output)
+		}
+	}
+}
+
+// SetLogger sets the StakingHub logger
+func (sh *StakingHub) SetLogger(logger hclog.Logger) {
+	sh.Logger = logger.Named("staking-hub")
+}
+
+// CloseStakingHub stops the writeback process
+func (sh *StakingHub) CloseStakingHub() {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	// Alert the closing channel
+	sh.CloseCh <- struct{}{}
+
+	close(sh.CloseCh)
+}
+
+// PendingEvent contains useful information about a staking / unstaking event
+type PendingEvent struct {
+	Number    int64
+	Address   types.Address
+	Value     *big.Int
+	EventType StakingEventType
+}
+
+// Compare checks if the two events match
+func (pe *PendingEvent) Compare(event PendingEvent) bool {
+	if pe.EventType == event.EventType &&
+		pe.Address.String() == event.Address.String() &&
+		pe.Value.Cmp(event.Value) == 0 &&
+		pe.Number == event.Number {
+		return true
+	}
+
+	return false
+}
+
+// AddPendingEvent pushes an event to the event queue
+func (sh *StakingHub) AddPendingEvent(event PendingEvent) {
+	sh.EventQueueMutex.Lock()
+	defer sh.EventQueueMutex.Unlock()
+
+	// Check to avoid double addition
+	if !sh.hasEvent(event) {
+		sh.EventQueue = append(sh.EventQueue, event)
+	}
+}
+
+// RemovePendingEvent removes the pending event from the event queue if it exists.
+// Returns a boolean flag indicating if an event was found and removed
+func (sh *StakingHub) RemovePendingEvent(event PendingEvent) bool {
+	sh.EventQueueMutex.Lock()
+	defer sh.EventQueueMutex.Unlock()
+
+	foundIndx := -1
+	foundFlag := false
+	for indx, el := range sh.EventQueue {
+		if el.Compare(event) {
+			foundIndx = indx
+		}
+	}
+
+	if foundIndx >= 0 {
+		sh.EventQueue = append(sh.EventQueue[:foundIndx], sh.EventQueue[foundIndx+1:]...)
+		foundFlag = true
+	}
+
+	return foundFlag
```

Comment:
I think you can do without `foundFlag`
```suggestion
	if foundIndx >= 0 {
		sh.EventQueue = append(sh.EventQueue[:foundIndx], sh.EventQueue[foundIndx+1:]...)
		return true
	}

	return false
```

Reply:
You're right.

Removing now 👍